language/src/main.rs

```diff
@@ -65,93 +68,207 @@ impl HacspecErrorEmitter for Session {
+                span,
             }
-        );
+        }
+    };
+
+    /////////////////////////////////
+    // Start of actual translation //
+    /////////////////////////////////
+
+    let external_data = |imported_crates: &Vec<rustspec::Spanned<String>>| {
+        queries.global_ctxt().unwrap().peek_mut().enter(|tcx| {
+            hir_to_rustspec::retrieve_external_data(&compiler.session(), &tcx, imported_crates)
+        })
+    };
+
+    let krate = match ast_to_rustspec::translate(&compiler.session(), &krate, &external_data) {
+        Ok(krate) => krate,
+        Err(_) => {
+            compiler
+                .session()
+                .err("unable to translate to Hacspec due to out-of-language errors");
+            return Compilation::Stop;
+        }
+    };
 
-        match &self.output_file {
-            None => return Compilation::Stop,
-            Some(file) => match Path::new(file).extension().and_then(OsStr::to_str).unwrap() {
+    let krate = match name_resolution::resolve_crate(
+        &compiler.session(),
+        krate,
+        &external_data,
+        new_top_ctx,
+    ) {
+        Ok(krate) => krate,
+        Err(_) => {
+            compiler
+                .session()
+                .err("found some Hacspec name resolution errors");
+            return Compilation::Stop;
+        }
+    };
+
+    let krate = match typechecker::typecheck_program(&compiler.session(), &krate, new_top_ctx) {
+        Ok(krate) => krate,
+        Err(_) => {
+            compiler
+                .session()
+                .err("found some Hacspec typechecking errors");
+            return Compilation::Stop;
+        }
+    };
+    let imported_crates = name_resolution::get_imported_crates(&krate);
+    let imported_crates = imported_crates
+        .into_iter()
+        .filter(|(x, _)| x != "hacspec_lib")
+        .map(|(x, _)| x)
+        .collect::<Vec<_>>();
+    println!(
+        " > Successfully typechecked{}",
+        if imported_crates.len() == 0 {
+            ".".to_string()
+        } else {
+            format!(
+                ", assuming that the code in crates {} has also been Hacspec-typechecked",
+                imported_crates.iter().format(", ")
+            )
+        }
+    );
+
+    match &callback.output_directory {
+        None => return Compilation::Stop,
+        Some(file_str) => {
+            let original_file = Path::new(file_str);
+            let extension = &callback.output_type;
+
+            // Compute file name as output directory with crate local path (krate_path)
+            let oe = 
+                (original_file)
+                .join(Path::new(
+                    krate_path
+                        .clone()
+                        .to_title_case()
+                        .replace(" ", "_")
+                        .replace("Hacspec_", "")
+                        .as_str(),
+                ))
+                .with_extension(extension);
```

Comment:
I am not sure about the output file name format, it works for Coq, but I am not sure it is correct for the other backends.

Reply:
F* uses `.` as separator and doesn't drop the `"Hacspec."` (check the files in the fstar folder).